### PR TITLE
Added wait and retries for rate limit response code from Bitbucket server

### PR DESCRIPTION
**Pull request by shankarinece**
Added 5 seconds wait and retry the same API if Bitbucket responds with rate-limit response code (429). Relates to JENKINS-62479, 

### Your checklist for this pull request

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

**Comment by ryancurrah** [comment]
This is a really handy improvement that would be beneficial for everyone.

**Comment by Zauxst** [comment]
@bitwiseman , are you in charge of merging PRs?  
Can you take a look over this one? 

The rate API is quite important to be fixed especially in large organizations…
Also, Bitbucket does not have any response headers regarding RATE API as mentioned in #414

**Comment by tarioch** [comment]
If you have many open prs, #445 might be important for you as well, on our setup it reduced the number of api calls massively

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 1]
```diff
@@ -524,6 +527,15 @@ public boolean checkPathExists(@NonNull String branchOrHash, @NonNull String pat
             // https://support.atlassian.com/bitbucket-cloud/docs/use-bitbucket-rest-api-version-1/
         } else if (HttpStatus.SC_NOT_FOUND == status || HttpStatus.SC_UNAUTHORIZED == status) {
             return false;
+        } else if (API_RATE_LIMIT_CODE == status) {
+            LOGGER.fine("Bitbucket API rate limit reached, sleeping for 5 sec then retry...");
+            try {
+                Thread.sleep(API_RATE_LIMIT_WAIT_TIME_DEFAULT);
```
Comment (bitwiseman) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java:533]: Just waiting for 5 seconds will cause a lot of polling and spamming of the server. Please shift to some kind of progressive back off where each of retry waits longer (up to some reasonable limit. 

Alternative: see if they the headers of the 429 provide information about when to the limit will reset and sleep for something like that long.

**Review by bitwiseman — CHANGES_REQUESTED** [review]
This is great start! Please address the comments and add some tests of the behavior.

**Comment by bitwiseman** [comment]
See #414

**Comment by smirkybg** [comment]
Can we get some progress on this? I'm constantly hitting 429 on our Jenkins setup and we definitely welcome this rate limit progressive back off strategy.

**Comment by lifeofguenter** [comment]
Closing in favor of https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/581

**Pull request closed without merging**